### ISSUE TITLE
Add ability to cancel "signin submit" event

### ISSUE THREAD

**Issue by jamiegaines** [auth0/lock#1063]
**Enhancement**: I'd like the ability to cancel the "signin submit" event.

**Use Case** : We have implemented a self-hosted login screen that serves as an enterprise sign in solution for a variety of applications.  Everything works great except for the password grant scenario, which I understand the lock control doesn't currently support for login screens hosted external to Auth0.

To work around this, we've used the signin submit event to trap the condition where a user is trying to signin with a password.  We then pass the login off to the authjs library's client.login method.

This works great, except there seems to be no way to stop the signin submit event from propgating.  The net result is that the lock control sends our original login attempt on to Auth0, and we get redirected to the "Password login is disabled for clients using externally hosted login pages with oidc_conformant flag set." message.

If we were able to cancel that signin submit event from within the handler, we'd have a nice workaround in place for the password grant scenario.  

Thanks!

**Comment by hzalaz** [comment]
@jamiegaines I have a bit of a hard time understanding your issue/request

- can you elaborate more what options your users could use to authenticate?
- can you explain how you are configuring Lock? and also what version?
- Are you supporting only enterprise connections?

**Comment by jamiegaines** [comment]
Sure.  We support a wide variety of enterprise web applications.  Some are for internal use, some are B2B, and others are B2C.  Currently, each application has its own login screen, homebrew or ASP.NET security, etc.

We want to use Auth0 to abstract the authentication code away from all of our existing applications.  Our goal is that no application should have to implement its own login screen.  When a given application detects that the user requires authentication, it will redirect to a common accounts.mycompany.com web application that uses the lock control to handle logins.

The redirect contains various metadata in the query string that lets the signin application know how to render the lock control for a given application.  Each application has its own Client defined in Auth0, thus  each application can have different set of Connections defined.

All Clients - including this accounts.blah.com web app - are set to be OIDC compliant and we're using the Auth0 SSO setting.  Each application has a small library that it runs on startup to detect if a valid token is present.  If not, the application tries to login silently, hoping that a SSO login has already taken place.  If silent login fails, the application redirects to our common sign in web app.

With all of this in mind, the base configuration of our lock control looks like this....

(Note: we render the control with ASP.NET Core MVC, hence the Razor syntax

```
var lock = new Auth0Lock('@Model.ClientId',
      '@Model.Domain',
      {
        allowedConnections: [@Html.Raw(Model.IdentityProviderString)],
        allowSignUp: false,
        closable: false,
        auth: {
          SSO: true,
          redirectUrl: '@Model.CallbackUrl',
          responseType: 'code',
          params: {
            scope: 'openid @Model.TenantId',
            state: '@Model.State',
            nonce: '@Model.Nonce'
          }
        }
       // other stuff removed for clarity

      });
```

That Model.IdentityProviderString variable contains the list of connections specific to the Client for which we are trying to authenticate, and the connections can vary wildly from application to application.  Some only support our internal Azure AD, some support social logins, some support only password/database logins.

We are using the built-in Auth0 Database connection to house username/password data for our users that don't authenticate via one of the 3rd party OIDC providers.  

This is where the problem lies.  Our setup above works great for everything except the database login.  When a user enters a username and password in the lock control, then hits submit, we get an error from Auth0 that says "Password login is disabled for clients using externally hosted login pages with oidc_conformant flag set"

Some Googling around - [here](https://auth0.com/forum/t/password-login-now-disabled-for-externally-hosted-pages/5395/2) and [here](https://community.auth0.com/questions/4943/lock-and-oidc-conformant-clients-1) - indicates that the lock control currently doesn't support this scenario, but that it might at a future time.

The guidance in both of those posts above seems to be to use the authjs library and call the client.login method to do your authentication instead.  That's fine, but we still want our users to use the lock control, and not go down some other path in the event they want to use password logins.

This is where the "signin submit" event comes in to play.  I've hacked together an event handler that isolates the scenario where the user is attempting to make a password login, then passes that username/password data off to the authjs library instead.

That handler roughly looks like this....

```
    lock.on("signin submit", invokePasswordLogin);

    function invokePasswordLogin() {
      var emailInput = document.getElementsByName("email")[0];
      var pwdInput = document.getElementsByName("password")[0];

      // my hacky way to see if this is a password login attempt
      if (pwdInput) {
        console.log("This is a password grant.");
        console.log("Submitting client login request via authjs.");
        webAuth.client.login({
            realm: 'OurDatabaseConnectionName',
            username: emailInput.value,
            password: pwdInput.value
          },
          function(err, authResult) {
            console.log("Received response from authjs client.login request.");
            // Auth tokens in the result or an error
            if (err) {
              console.log("Login error");
              console.log(err);
            }

            if (authResult) {
              console.log("Login success");
            }

          });
      }
    }
```

This all works fine too.  I am able to call the webAuth.client.login method and receive the authResult for a valid login in the handler. At this point, I'd like to take the token, hand it back to my accounts.blah.com web app, and go on about my business.  

The problem is that I have no way to stop the "signin submit" event from propogating.  What happens is that the lock control submits the form to auth0, and before my code above can do anything, I've been redirected to Auth0 and am seeing the "Password login is disabled for clients using externally hosted login pages with oidc_conformant flag set." message.

What I'd like to see is something like this....

```
lock.on("signin submit", function(e) {
  var result = invokePasswordLogin();

  if(result === "success") {
    handlePasswordLoginSuccess();
  }
  else {
   handlePasswordLoginError();
  }

  // ********* This is the money line *****************
  e.stopPropogation();
  e.PreventDefault();
  e.JustPleaseStop();  // whatever it takes to cancel the form submit event.
);

```

In know you guys are using React, so passing the native event might not be the best option.  Perhaps if the signin submit handler could return true/false and the lock control could decide on its own to stop the event or not based upon the result.

Anyway, we are thisclose to getting this all working, and this one issue is holding us up right now.

Sorry this is so long-winded, but I hope it does a better job of explaining what we're trying to do.

Cheers.

**Comment by hzalaz** [comment]
@jamiegaines cool now I get the full picture and I believe we are working on something to address your use case (db login in your site with lock) without the need to override the login submit (which we don't recommend). 

Just to confirm by 

> All Clients - including this accounts.blah.com web app - are set to be OIDC compliant 

This is the flag in Auth0 Dashboard right?

**Comment by jamiegaines** [comment]
Yes, we set the flag in the Advanced Settings section.

That's exciting you are working on incorporating the password login into the lock control.  It will be nice to get my hack out of there.  :)

In the interim, I ended up with another hack. After the lock control loads - be nice to have an event for that too :) - I get a reference to the submit button via jQuery and replace it with my own, if it exists.  Then I have a hook in to handling the event myself, or letting it pass through to the form submit.

Anyway, thanks for taking the time to respond.  You guys are doing a great job.

Cheers.
